tt_eager/tensor/tensor.cpp

```diff
@@ -107,6 +107,9 @@ void Tensor::deallocate(bool force) {
                         std::visit([](auto&& buffer) { buffer.reset(); }, storage.buffer);
                     }
                 } else if constexpr (std::is_same_v<T, DeviceStorage>) {
+                    if (this->workers.at(0) == nullptr or not this->workers.at(0)->is_initialized()) {
```

Comment:
@aliuTT @tt-asaigal Did we do some more investigation into this?

Reply:
Only doable on the DevicePool changes. This is needed on main, I want to merge this in separately because some models people/CI are blocked.

Reply:
Which models are these? I thought this was only potentially affecting workloads after device close, but those tests were disabled.

Reply:
New models under dev. I was given a repro case with Llama3, this fix helped them. I think they need this in main for those model tests to get checked into CI

Reply:
I'd like to strongly push for some kind of unit test. Is there anyway to repro this with a smaller test and add that along with this PR? Otherwise, I'm concerned we're adding a patch without fully understanding what's going on. It'll be helpful having a regression test so that your next PR can also make sure we are still regressing on this.

Reply:
@kevinmiTT11 is it possible to get a small unit test that can repro the segfault you were seeing with Llama3?

Reply:
@mtairum Can you provide a unit test where this change was making a difference? I heard it was helping with some of your model tests.

Reply:
I don't think I've tried this change before.

I remember trying the garbage collector change from a week or two ago and that was helpful in dealing with segfault when closing devices after a test fails.

I try to replicate some of our tests assert failing but none of them segfaults on tt-metal tip from yesterday.

Reply:
There is a new model seg fault, maybe we can take a look at the call trace at least?

https://github.com/tenstorrent/tt-metal/issues/9268